File: server/services/audit_service.py
"""Audit service for hash chain management."""
from datetime import datetime, timezone
from typing import Any

from server.crypto.hashing import compute_hash, verify_chain as crypto_verify_chain


class AuditService:
    """Service for managing the immutable audit hash chain."""

    async def record(
        self,
        db_session,
        agent_did: str,
        transaction_id: str,
        amount: int,
        status: str,
        timestamp: str,
    ) -> str:
        """Record a transaction in the audit hash chain.

        Args:
            db_session: Async SQLAlchemy session
            agent_did: DID of the agent
            transaction_id: Unique transaction identifier
            amount: Transaction amount
            status: Transaction status
            timestamp: ISO 8601 timestamp

        Returns:
            str: Hash of the new audit log entry
        """
        from server.models.audit_log import AuditLog
        from sqlalchemy import select, func

        # Get the last audit log entry for this agent
        result = await db_session.execute(
            select(AuditLog)
            .where(AuditLog.agent_did == agent_did)
            .order_by(AuditLog.index_num.desc())
            .limit(1)
        )
        last_entry = result.scalar_one_or_none()

        if last_entry is None:
            prev_hash = "0" * 64
            index = 0
        else:
            prev_hash = last_entry.hash
            index = last_entry.index_num + 1

        # Compute hash for new entry
        entry_hash = compute_hash(
            index=index,
            transaction_id=transaction_id,
            amount=amount,
            status=status,
            timestamp=timestamp,
            prev_hash=prev_hash,
        )

        # Save to database
        audit_log = AuditLog(
            index_num=index,
            agent_did=agent_did,
            transaction_id=transaction_id,
            amount=amount,
            status=status,
            timestamp=timestamp,
            prev_hash=prev_hash,
            hash=entry_hash,
        )
        db_session.add(audit_log)

        return entry_hash

    async def get_chain(self, db_session, agent_did: str) -> dict[str, Any]:
        """Get the complete audit chain for an agent.

        Args:
            db_session: Async SQLAlchemy session
            agent_did: DID of the agent

        Returns:
            dict with chain entries and validation status
        """
        from server.models.audit_log import AuditLog
        from sqlalchemy import select

        result = await db_session.execute(
            select(AuditLog)
            .where(AuditLog.agent_did == agent_did)
            .order_by(AuditLog.index_num.asc())
        )
        entries = result.scalars().all()

        chain = []
        for entry in entries:
            chain.append({
                "index": entry.index_num,
                "transaction_id": entry.transaction_id,
                "amount": entry.amount,
                "status": entry.status,
                "timestamp": entry.timestamp,
                "prev_hash": entry.prev_hash,
                "hash": entry.hash,
            })

        chain_valid = crypto_verify_chain(chain) if chain else True

        total = len(chain)
        succeeded = sum(1 for e in chain if e["status"] == "succeeded")
        success_rate = succeeded / total if total > 0 else 1.0

        return {
            "agent_did": agent_did,
            "chain": chain,
            "chain_valid": chain_valid,
            "total_transactions": total,
            "success_rate": success_rate,
        }

    async def verify(self, db_session, agent_did: str) -> dict[str, Any]:
        """Verify the integrity of an agent's audit hash chain.

        Args:
            db_session: Async SQLAlchemy session
            agent_did: DID of the agent

        Returns:
            dict with verification result
        """
        from server.models.audit_log import AuditLog
        from sqlalchemy import select

        result = await db_session.execute(
            select(AuditLog)
            .where(AuditLog.agent_did == agent_did)
            .order_by(AuditLog.index_num.asc())
        )
        entries = result.scalars().all()

        chain = []
        for entry in entries:
            chain.append({
                "index": entry.index_num,
                "transaction_id": entry.transaction_id,
                "amount": entry.amount,
                "status": entry.status,
                "timestamp": entry.timestamp,
                "prev_hash": entry.prev_hash,
                "hash": entry.hash,
            })

        chain_valid = crypto_verify_chain(chain) if chain else True
        verified_at = datetime.now(timezone.utc).isoformat()

        return {
            "agent_did": agent_did,
            "chain_valid": chain_valid,
            "total_entries": len(chain),
            "verified_at": verified_at,
        }


audit_service = AuditService()

File: server-rust/src/services/audit_service.rs
use chrono::{DateTime, Utc};
use sqlx::PgPool;

use crate::crypto::hashing::{compute_hash, verify_chain, AuditEntry, GENESIS_PREV_HASH};
use crate::error::{AppError, AppResult};

pub async fn record(
    db: &PgPool,
    agent_did: &str,
    transaction_id: &str,
    amount: i64,
    status: &str,
    timestamp: DateTime<Utc>,
) -> AppResult<String> {
    // Get the last audit log entry for this agent
    let last = sqlx::query_as::<_, (i64, String)>(
        "SELECT index_num, hash FROM audit_logs WHERE agent_did = $1 ORDER BY index_num DESC LIMIT 1",
    )
    .bind(agent_did)
    .fetch_optional(db)
    .await?;

    let (index, prev_hash) = match last {
        Some((last_index, last_hash)) => (last_index + 1, last_hash),
        None => (0, GENESIS_PREV_HASH.to_string()),
    };

    // The timestamp string must match the Python isoformat() output
    let timestamp_str = timestamp.to_rfc3339_opts(chrono::SecondsFormat::Micros, true);

    let entry_hash = compute_hash(index, transaction_id, amount, status, &timestamp_str, &prev_hash);

    sqlx::query(
        r#"
        INSERT INTO audit_logs (index_num, agent_did, transaction_id, amount, status, timestamp, prev_hash, hash)
        VALUES ($1, $2, $3, $4, $5, $6, $7, $8)
        "#,
    )
    .bind(index)
    .bind(agent_did)
    .bind(transaction_id)
    .bind(amount)
    .bind(status)
    .bind(timestamp)
    .bind(&prev_hash)
    .bind(&entry_hash)
    .execute(db)
    .await?;

    Ok(entry_hash)
}

pub async fn get_chain(db: &PgPool, agent_did: &str) -> AppResult<serde_json::Value> {
    let rows = sqlx::query_as::<_, (i64, String, i64, String, DateTime<Utc>, String, String)>(
        r#"
        SELECT index_num, transaction_id, amount, status, timestamp, prev_hash, hash
        FROM audit_logs
        WHERE agent_did = $1
        ORDER BY index_num ASC
        "#,
    )
    .bind(agent_did)
    .fetch_all(db)
    .await?;

    let mut entries: Vec<AuditEntry> = Vec::new();
    let mut chain_json: Vec<serde_json::Value> = Vec::new();

    for (index_num, transaction_id, amount, status, timestamp, prev_hash, hash) in &rows {
        let timestamp_str = timestamp.to_rfc3339_opts(chrono::SecondsFormat::Micros, true);
        chain_json.push(serde_json::json!({
            "index": index_num,
            "transaction_id": transaction_id,
            "amount": amount,
            "status": status,
            "timestamp": timestamp_str,
            "prev_hash": prev_hash,
            "hash": hash,
        }));
        entries.push(AuditEntry {
            index: *index_num,
            transaction_id: transaction_id.clone(),
            amount: *amount,
            status: status.clone(),
            timestamp: timestamp_str,
            prev_hash: prev_hash.clone(),
            hash: hash.clone(),
        });
    }

    let chain_valid = if entries.is_empty() {
        true
    } else {
        verify_chain(&entries)
    };

    let total = entries.len();
    let succeeded = entries.iter().filter(|e| e.status == "succeeded").count();
    let success_rate = if total > 0 {
        succeeded as f64 / total as f64
    } else {
        1.0
    };

    Ok(serde_json::json!({
        "agent_did": agent_did,
        "chain": chain_json,
        "chain_valid": chain_valid,
        "total_transactions": total,
        "success_rate": success_rate,
    }))
}

pub async fn verify(db: &PgPool, agent_did: &str) -> AppResult<serde_json::Value> {
    let rows = sqlx::query_as::<_, (i64, String, i64, String, DateTime<Utc>, String, String)>(
        r#"
        SELECT index_num, transaction_id, amount, status, timestamp, prev_hash, hash
        FROM audit_logs
        WHERE agent_did = $1
        ORDER BY index_num ASC
        "#,
    )
    .bind(agent_did)
    .fetch_all(db)
    .await?;

    let entries: Vec<AuditEntry> = rows
        .iter()
        .map(|(index_num, transaction_id, amount, status, timestamp, prev_hash, hash)| {
            let timestamp_str = timestamp.to_rfc3339_opts(chrono::SecondsFormat::Micros, true);
            AuditEntry {
                index: *index_num,
                transaction_id: transaction_id.clone(),
                amount: *amount,
                status: status.clone(),
                timestamp: timestamp_str,
                prev_hash: prev_hash.clone(),
                hash: hash.clone(),
            }
        })
        .collect();

    let chain_valid = if entries.is_empty() {
        true
    } else {
        verify_chain(&entries)
    };

    let verified_at = Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Micros, true);

    Ok(serde_json::json!({
        "agent_did": agent_did,
        "chain_valid": chain_valid,
        "total_entries": entries.len(),
        "verified_at": verified_at,
    }))
}
